crates/matrix-sdk-ui/src/timeline/polls.rs

```diff
@@ -0,0 +1,195 @@
+/// Polls module.
+///
+/// This module handles rendering of MSC3381 polls in the timeline.
+use imbl::HashMap;
+use ruma::{
+    events::{
+        poll::{
+            compile_unstable_poll_results,
+            start::PollKind,
+            unstable_response::{
+                OriginalSyncUnstablePollResponseEvent, UnstablePollResponseEventContent,
+            },
+            unstable_start::{
+                UnstablePollStartContentBlock, UnstablePollStartEventContent,
+                UnstablePollStartEventContentWithoutRelation,
+            },
+        },
+        MessageLikeUnsigned,
+    },
+    server_name, EventId, MilliSecondsSinceUnixEpoch, OwnedEventId, OwnedUserId,
+};
+
+/// Holds the state of a poll.
+///
+/// This struct should be created for each poll start event handled and then
+/// updated whenever handling any poll response or poll end event that relates
+/// to the same poll start event.
+#[derive(Clone, Debug)]
+pub struct PollState {
+    pub(super) start_event_content: UnstablePollStartEventContent,
+    pub(super) response_events: Vec<ResponseEvent>,
+    pub(super) end_event_timestamp: Option<MilliSecondsSinceUnixEpoch>,
+}
+
+#[derive(Clone, Debug)]
+pub(super) struct ResponseEvent {
+    pub(super) sender: OwnedUserId,
+    pub(super) timestamp: MilliSecondsSinceUnixEpoch,
+    pub(super) content: UnstablePollResponseEventContent,
+}
+
+impl PollState {
+    pub(super) fn new(content: UnstablePollStartEventContent) -> Self {
+        Self { start_event_content: content, response_events: vec![], end_event_timestamp: None }
+    }
+
+    pub(super) fn edit(
+        &self,
+        replacement: &UnstablePollStartEventContentWithoutRelation,
+    ) -> Result<Self, ()> {
+        if self.response_events.is_empty() && self.end_event_timestamp.is_none() {
+            let mut clone = self.clone();
+            clone.start_event_content.poll_start = replacement.poll_start.clone();
+            clone.start_event_content.text = replacement.text.clone();
+            Ok(clone)
+        } else {
+            Err(())
+        }
+    }
+
+    pub(super) fn add_response(
+        &self,
+        sender: &OwnedUserId,
+        timestamp: &MilliSecondsSinceUnixEpoch,
+        content: &UnstablePollResponseEventContent,
+    ) -> Self {
+        let mut clone = self.clone();
+        clone.response_events.push(ResponseEvent {
+            sender: sender.clone(),
+            timestamp: timestamp.clone(),
+            content: content.clone(),
+        });
+        clone
+    }
+
+    pub(super) fn end(&self, timestamp: &MilliSecondsSinceUnixEpoch) -> Self {
+        let mut clone = self.clone();
+        clone.end_event_timestamp = Some(timestamp.clone());
+        clone
+    }
+
+    pub fn fallback_text(&self) -> Option<String> {
+        self.start_event_content.text.clone()
+    }
+
+    pub fn results(&self) -> PollResult {
```

Comment:
This would be the same as running the logic in `PollState.result()` every time `PollState.add_response()` or `PollState.end()` are called.

Therefore the answer depends on which methods are called more times on average usage:

`PollState.result()` (which is only called from in `matrix_sdk_ffi::timeline::TimelineItemContent.kind()` )

OR

`PollState.add_response()` and `PollState.end()` (which are called from `matrix_sdk_ui::timeline::event_handler::TimelineEventHandler.handle_event()`)

?